File: demo-app/app/src/main/java/com/launchkey/android/authenticator/demo/ui/adapter/DemoDevicesAdapter.java
package com.launchkey.android.authenticator.demo.ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.TextView;

import com.launchkey.android.authenticator.sdk.device.Device;
import com.launchkey.android.authenticator.demo.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by armando on 7/11/16.
 */
public class DemoDevicesAdapter extends BaseAdapter {

    private Context mContext;
    private List<Device> mDevices = new ArrayList<>();
    private AdapterView.OnItemClickListener mItemClickListener;
    private View.OnClickListener mInternalClickListener;

    public DemoDevicesAdapter(Context c, List<Device> devices, AdapterView.OnItemClickListener l) {
        mContext = c;
        mDevices = devices;
        mItemClickListener = l;

        mInternalClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (v != null && v.getTag() != null) {
                    int position = (int) v.getTag();

                    if (mItemClickListener != null) {
                        mItemClickListener.onItemClick(null, v, position, position);
                    }
                }
            }
        };
    }

    @Override
    public int getCount() {
        return mDevices.size();
    }

    @Override
    public Device getItem(int position) {
        return mDevices.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        View v = convertView;

        if (v == null) {
            v = LayoutInflater
                    .from(mContext)
                    .inflate(R.layout.demo_devices_item, parent, false);
        }

        Device d = getItem(position);

        TextView currentDevice = (TextView) v.findViewById(R.id.demo_devices_item_currentdevice);
        currentDevice.setVisibility(position == 0 ? View.VISIBLE : View.GONE);

        TextView name = (TextView) v.findViewById(R.id.demo_devices_item_name);
        name.setText(d.getName());

        TextView status = (TextView) v.findViewById(R.id.demo_devices_item_status);
        status.setText(d.getType());

        Button button = (Button) v.findViewById(R.id.demo_devices_item_button);
        button.setOnClickListener(mInternalClickListener);
        button.setTag(position);

        return v;
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/adapter/DemoDevicesAdapter.kt
package com.launchkey.android.authenticator.demo.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.TextView
import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.sdk.device.Device
import java.util.*

/**
 * Created by armando on 7/11/16.
 */
class DemoDevicesAdapter(private val mContext: Context, devices: List<Device>, private val mItemClickListener: AdapterView.OnItemClickListener?) : BaseAdapter() {
    private var mDevices: List<Device> = ArrayList()
    private val mInternalClickListener: View.OnClickListener

    init {
        mDevices = devices

        mInternalClickListener = View.OnClickListener { v ->
            if (v != null && v.tag != null) {
                val position = v.tag as Int

                mItemClickListener?.onItemClick(null, v, position, position.toLong())
            }
        }
    }

    override fun getCount(): Int {
        return mDevices.size
    }

    override fun getItem(position: Int): Device {
        return mDevices[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {

        var v: View? = convertView

        if (v == null) {
            v = LayoutInflater
                    .from(mContext)
                    .inflate(R.layout.demo_devices_item, parent, false)
        }

        val d = getItem(position)

        val currentDevice = v!!.findViewById<View>(R.id.demo_devices_item_currentdevice) as TextView
        currentDevice.visibility = if (position == 0) View.VISIBLE else View.GONE

        val name = v.findViewById<View>(R.id.demo_devices_item_name) as TextView
        name.text = d.name

        val status = v.findViewById<View>(R.id.demo_devices_item_status) as TextView
        status.text = d.type

        val button = v.findViewById<View>(R.id.demo_devices_item_button) as Button
        button.setOnClickListener(mInternalClickListener)
        button.tag = position

        return v
    }
}
